File: c++/src/lib/dmp_lib/src/DMP/WSoG.cpp
#include <dmp_lib/DMP/WSoG.h>

namespace as64_
{

namespace dmp_
{

WSoG::WSoG(int N_kernels, std::function<double(double)> shapeAttrGatingFun, double kernel_std_scaling)
{
  this->N_kernels = N_kernels;
  this->shapeAttrGatingFun = shapeAttrGatingFun;

  this->zero_tol = 1e-30; // realmin;

  this->w = arma::vec().zeros(N_kernels);
  this->c = arma::linspace<arma::vec>(0,N_kernels-1, N_kernels)/(N_kernels-1);
  this->h.resize(N_kernels);
  for (int i=0; i<N_kernels-1; i++) this->h(i) = 1 / std::pow(kernel_std_scaling*(this->c(i+1)-this->c(i)),2);
  this->h(N_kernels-1) = this->h(N_kernels-2);
}


int WSoG::getNumOfKernels() const
{
  return this->w.size();
}


void WSoG::train(dmp_::TRAIN_METHOD train_method, const arma::rowvec &x, const arma::rowvec &Fd, double *train_error)
{
  int n_data = x.size();

  arma::rowvec s(n_data);
  arma::mat Psi(this->N_kernels, n_data);
  for (int j=0; j<n_data; j++)
  {
    s(j) = this->shapeAttrGatingFun(x(j));
    Psi.col(j) = this->kernelFunction(x(j));
  }

  if (train_method == dmp_::LWR) this->w = dmp_::localWeightRegress(Psi, s, Fd, this->zero_tol);
  else if (train_method == dmp_::LS) this->w = dmp_::leastSquares(Psi, s, Fd, this->zero_tol);
  else throw std::runtime_error("[WSoG::train]: Unsopported training method...");

  if (train_error)
  {
    arma::rowvec F(n_data);
    for (int j=0; j<n_data; j++) F(j) = this->output(x(j));
    *train_error = arma::norm(F-Fd)/n_data;
  }

};


arma::vec WSoG::kernelFunction(double x) const
{
  arma::vec psi = arma::exp(-this->h % (arma::pow(x-this->c,2)));
  return psi;
}


double WSoG::output(double x) const
{
  arma::vec psi = this->kernelFunction(x);
  double f = this->shapeAttrGatingFun(x) * dot(psi,this->w) / (sum(psi)+this->zero_tol); // add 'zero_tol' to avoid numerical issues
  return f;
}

} // namespace dmp_

} // namespace as64_

File: matlab/utils/lib/DMP_lib/WSoG.m
%% WSoG class
%  Weighted sum of Guassians.
%

classdef WSoG < handle
    
    properties (Constant)
          
          %% enum TRAIN_METHOD
          LWR = 201;
          LS = 203;

    end
       
    properties
        N_kernels % number of kernels (basis functions)

        w % N_kernels x 1 vector with the kernels' weights
        c % N_kernels x 1 vector with the kernels' centers
        h % N_kernels x 1 vector with the kernels' inverse width

        shapeAttrGatingFun % Pointer to gating function that ensures the output of WSoG decays to zero at the end (i.e. when the phase variable reaches 1.0)
        
        zero_tol % small value used to avoid divisions with very small numbers
        
    end

    methods
        %% Weighted Sum of Gaussians constructor.
        %  @param[in] N_kernels: The number of kernels.
        %  @param[in] shapeAttrGatingFun: Pointer to gating function that ensures the output of WSoG decays
        %                                 to zero at the end (i.e. when the phase variable reaches 1.0).
        %  @param[in] kernel_std_scaling: Scaling of the kernel's std. (optional, default=1.0)
        %
        function this = WSoG(N_kernels, shapeAttrGatingFun, kernel_std_scaling) %, N_kernels, s_gat_ptr

            if (nargin < 3), kernel_std_scaling = 1.0; end
            
            this.N_kernels = N_kernels;
            this.shapeAttrGatingFun = shapeAttrGatingFun;
            
            this.zero_tol = 1e-30; %realmin;

            this.w = zeros(this.N_kernels,1);
            this.c = ((1:this.N_kernels)-1)'/(this.N_kernels-1);
            this.h = 1./(kernel_std_scaling*(this.c(2:end)-this.c(1:end-1))).^2;
            this.h = [this.h; this.h(end)];
            
        end

        %% Returns the number of kernels.
        %  @return The number of kernels.
        %
        function n_ker = getNumOfKernels(this)
            
            n_ker = length(this.w);
            
        end
        
        
        %% Trains the WSoG.
        %  @param[in] train_method: The training method (see dmp_::TRAIN_METHOD enum).
        %  @param[in] x: Row vector with the timestamps of the training data points. Must take values in [0 1].
        %  @param[in] Fd: Row vector with the desired values.
        %  @param[in] train_error: Optinal pointer to return the training error as norm(F-Fd)/n_data.
        %
        function [train_error, F] = train(this, train_method, x, Fd)

            n_data = length(x);

            s = zeros(1, n_data);
            Psi = zeros(this.N_kernels, n_data);
            for i=1:n_data
                s(i) = this.shapeAttrGatingFun(x(i));
                Psi(:,i) = this.kernelFunction(x(i));
            end

            if (train_method == WSoG.LWR), this.w = LWR(Psi, s, Fd, this.zero_tol);
            elseif (train_method == WSoG.LS), this.w = leastSquares(Psi, s, Fd, this.zero_tol);
            else, error('[WSoG::train]: Unsopported training method...');
            end

            if (nargout > 0)
                F = zeros(size(Fd));
                for i=1:size(F,2)
                    F(i) = this.output(x(i));
                end
                train_error = norm(F-Fd)/length(F);
            end

        end

        
        %% Returns a column vector with the values of the kernel functions.
        %  @param[in] x: The phase variable.
        %  @return: Column vector with the values of the kernel functions.
        %
        function psi = kernelFunction(this,x)

            n = length(x);
            psi = zeros(this.N_kernels, n);

            for j=1:n
                psi(:,j) = exp(-this.h.*((x(j)-this.c).^2));
            end 

        end
         
        
        %% Returns the normalized weighted sum of the Gaussians for the given phase variable (time instant).
        %  @param[in] x: The phase variable.
        %  @param[out] f: The normalized weighted sum of the Gaussians.
        %
        function f = output(this,x)

            Psi = this.kernelFunction(x);
    
            f = this.shapeAttrGatingFun(x) * dot(Psi,this.w) / (sum(Psi)+this.zero_tol); % add 'zero_tol' to avoid numerical issues

        end

    end
end
